### PR TITLE
Drop parens around Pexp_fun in argument position

### PR DESCRIPTION
**Pull request by IwanKaramazow**
Before
```reason
  switch (action, state) {
  | Click =>
    ReasonReact.UpdateWithSideEffects(
      {...state, click: click + 1},
      (
        self => {
          let _ = 1;
          apply(bar);
        }
      )
    )
```
After
```reason
  switch (action, state) {
  | Click =>
    ReasonReact.UpdateWithSideEffects(
      {...state, click: click + 1},
      self => {
        let _ = 1;
        apply(bar);
      }
    )
```

### CODE REVIEW

File: formatTest/unit_tests/expected_output/syntax.re [hunk 1]
```diff
@@ -1284,10 +1284,10 @@ Thing.map(
   foo,
   bar,
   baz,
-  [@attr]
-  ((abc, z) => MyModuleBlah.toList(argument)),
-  [@attr]
-  ((abc, z) => MyModuleBlah.toList(argument)),
+  [@attr] (abc, z) =>
```
Comment (jordwalke) [inline formatTest/unit_tests/expected_output/syntax.re]: This formatting is a little weird. We might consider improving it next time we're in this part of the formatter.

File: src/reason-parser/reason_pprint_ast.ml [hunk 2]
```diff
@@ -3600,13 +3600,13 @@ let printer = object(self:'self)
    * TODO: in the future we should probably use the type ruleCategory
    * to 'automatically' ensure the validity of a constraint expr with parens...
    *)
-  method unparseConstraintExpr ?(ensureExpr=false) e =
+  method unparseProtectedExpr ?(forceParens=false) e =
```
Comment (jordwalke) [inline src/reason-parser/reason_pprint_ast.ml]: The name here "constraint expr" referred to type constraint IIRC. What does "protected" mean?

Comment (IwanKaramazow) [reply src/reason-parser/reason_pprint_ast.ml]: I'll submit a PR with some comments.

File: formatTest/unit_tests/expected_output/syntax.re (hunk 1 above)
Comment (IwanKaramazow) [reply formatTest/unit_tests/expected_output/syntax.re]: How would you like to see this?

**Pull request merged**